### ISSUE TITLE
Whitespace and comments converted to (?:) when using x flag

### ISSUE THREAD

**Issue by lencioni** [slevithan/xregexp#130]
I'm not sure if this is as intended or not, but I noticed that when I am using the `x` flag, whitespace and comments are converted to `(?:)`. For example, this causes a regex like:

```js
xRegExp(' foo # comment', 'x');
```

to be transpiled to 

```js
/(?:)foo(?:)(?:)/
```

In this simple case it isn't that much, but for larger regexes like:

```js
xRegExp(`
  \\s*                     # preceding whitespace
  (?:
    (?<quote>['"])use\\sstrict\\k<quote>;? # 'use strict';
    |
    //.*                   # single-line comment
    |
    /*                     # open multi-line comment
    (?:\n|.)*?             # inside of multi-line comment
    \\*/                   # close multi-line comment
  )?                       # ? b/c we want to match on only whitespace
  \n                       # end of line
  `, 'xs'
);
```

we end up with

```js
/(?:)\s*(?:)(?:)(?:)(?:(?:)(['"])use\sstrict\1;?(?:)(?:)(?:)|(?:)\/\/[\s\S]*(?:)(?:)(?:)|(?:)\/*(?:)(?:)(?:)(?:(?:)|[\s\S])*?(?:)(?:)(?:)\*\/(?:)(?:)(?:))?(?:)(?:)(?:)(?:)/
```

which seems pretty wacky to me. Without the extra `(?:)` groups, we would have:

```js
/\s*(?:(['"])use\sstrict\1;?|\/\/[\s\S]*|\/*(?:|[\s\S])*?\*\/)?/
```

Is there any reason for the empty `(?:)` groups? I also wonder if #129 is related to this.

**Comment by lencioni** [comment]
I think I see the code that does this. It looks like it adds this group to "keep tokens separated unless the following token is a quantifier". I don't think I understand why they need to be kept separate in this case. And, if they do, it seems like only one would be necessary instead of 3.

**Comment by slevithan** [comment]
> Is there any reason for the empty (?:) groups?

Tokens separated by whitespace or comments can change their meaning if you make them adjacent. E.g. `\1 1` should match backreference 1 followed by a literal 1. If they appeared together as `\11`, this would match backreference 11 instead. Similarly `(?<name>)\k<name>1` must be transpiled to `()\1(?:)1`. And in other cases, this helps preserve invalid syntax as errors.

You can use e.g. `XRegExp('').xregexp.source` to get the original source rather than the transpiled source with empty groups inserted.

There used to be a cleanup routine that removed unneeded `(?:)`s, but I recently deleted it due to complexity (there are a few complex edge cases) and for perf concerns. I'll leave this issue open, because it should be easy/fast to just replace all cases of `(?:)(?:)` with `(?:)`, rather than trying to remove all unneeded empty groups as XRegExp previously did.

**Comment by lencioni** [comment]
Ah yes, that makes sense. Thanks for the explanation! Perhaps a short note about why this is done could be added to the comments for future explorers?

**Comment by slevithan** [comment]
Closed by commit 4ba2362b4119f8b32418d0c59ab58b457e773bb7.
